tools/bridge/tests/test_transfer_fetcher.py

```diff
@@ -1,23 +1,235 @@
+
+
+def test_fetch_token_transfer_events_in_range_ignore_transfers_not_to_bridge(
+    transfer_fetcher, token_transfer_event_queue, transfer_tokens_to, w3_foreign
+):
+    assert token_transfer_event_queue.empty()
+
+    transfer_tokens_to("0xbB1046b0Fe450aA48DEafF6Fa474AdBf972840dD")
+    transfer_fetcher._fetch_token_transfer_events_in_range(
+        0, w3_foreign.eth.blockNumber
+    )
+
+    assert token_transfer_event_queue.empty()
+
+
+def test_fetch_token_transfer_events_in_range_invalid_range(
+    transfer_fetcher, token_transfer_event_queue, transfer_tokens_to, w3_foreign
+):
+    with pytest.raises(AssertionError):
+        transfer_fetcher._fetch_token_transfer_events_in_range(1, 0)
+
+
+def test_fetch_token_transfer_events_not_seen_ignore_reorg_depth_blocks(
+    transfer_fetcher,
+    token_transfer_event_queue,
+    transfer_tokens_to_foreign_bridge,
+    w3_foreign,
+    max_reorg_depth,
+):
+    assert token_transfer_event_queue.empty()
+
+    transfer_tokens_to_foreign_bridge()
+
+    assert w3_foreign.eth.blockNumber < max_reorg_depth
+
+    transfer_fetcher._fetch_token_transfer_events_not_seen()
+
+    assert token_transfer_event_queue.empty()
+
+
+def test_fetch_token_transfer_events_not_seen(
+    transfer_fetcher,
+    token_transfer_event_queue,
+    transfer_tokens_to_foreign_bridge,
+    tester_foreign,
+    max_reorg_depth,
+):
+    assert token_transfer_event_queue.empty()
+
+    transfer_tokens_to_foreign_bridge()
+    tester_foreign.mine_blocks(max_reorg_depth)
+    transfer_fetcher._fetch_token_transfer_events_not_seen()
+
+    assert token_transfer_event_queue.qsize() == 1
+
+
+def test_fetch_token_transfer_events_not_seen_handle_log_limit_exact_multiplicate(
+    transfer_fetcher,
+    token_transfer_event_queue,
+    transfer_tokens_to_foreign_bridge,
+    tester_foreign,
+    max_reorg_depth,
+    event_filter_fetch_limit,
+    w3_foreign,
+):
+    assert token_transfer_event_queue.empty()
+
+    transfer_count = event_filter_fetch_limit * 2
+
+    for i in range(transfer_count):
+        transfer_tokens_to_foreign_bridge()
+
+    tester_foreign.mine_blocks(max_reorg_depth)
+    transfer_fetcher._fetch_token_transfer_events_not_seen()
+
+    assert token_transfer_event_queue.qsize() == transfer_count
+
+
+def test_fetch_token_transfer_events_not_seen_handle_log_limit_not_exact_multiplicate(
+    transfer_fetcher,
+    token_transfer_event_queue,
+    transfer_tokens_to_foreign_bridge,
+    tester_foreign,
+    max_reorg_depth,
+    event_filter_fetch_limit,
+):
+    assert token_transfer_event_queue.empty()
+
+    transfer_count = event_filter_fetch_limit + 1
+
+    for i in range(transfer_count):
+        transfer_tokens_to_foreign_bridge()
+
+    tester_foreign.mine_blocks(max_reorg_depth)
+    transfer_fetcher._fetch_token_transfer_events_not_seen()
+
+    assert token_transfer_event_queue.qsize() == transfer_count
+
+
+# TODO: test worker when async framework integrated
```

Comment:
leftover?

Reply:
Nope. Must be still tested, right?

Reply:
Then I don't get the comment. Do you mean you want to write more tests that test the class when run by gevent?

Reply:
I must somehow make sure that the `fetch_transfer_events` is working. Therefore yes, I must run it with `gevent`. Do you think different?

Reply:
No, but why can't you do this already but have to wait until "async framework integrated"?